### ISSUE TITLE
Find a way to detect page loads initiated by the loaded page

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/e2e-testing#25]
In #23 we found that it is possible to use a child window instead of an iframe, so we circumvented #3 . Now we need to investigate how it is possible to detect page loads initiated by the loaded page. A possible solution could be to listen to add an event handler to the unload event.

**Comment by inf3rno** [comment]
This is relatively easy. We need to add the unload event handler. So we will detect when the new page is loaded. We could support history api changes etc too, but not now.

**Comment by inf3rno** [comment]
Object.`__defineSetter__`

	var o = {};
	o.__defineSetter__('value', function(val) { this.anotherValue = val; });
	o.value = 5;
	console.log(o.value); // undefined
	console.log(o.anotherValue); // 5
	



mutation events

	window.addEventListener("DOMAttrModified", function (event) {
	  // might fire on document change, who knows
	  // I don't have high hopes about this
	}, false);
	
	
	/*DOMNodeInserted
	DOMNodeInsertedIntoDocument
	DOMNodeRemoved
	DOMNodeRemovedFromDocument
	DOMSubtreeModified*/
	

mutation observers

	var targetNode = document.getElementById('some-id');
	var config = { attributes: true, childList: true, subtree: true };
	var callback = function(mutationsList) {
		for(var mutation of mutationsList) {
			if (mutation.type == 'childList') {
				console.log('A child node has been added or removed.');
			}
			else if (mutation.type == 'attributes') {
				console.log('The ' + mutation.attributeName + ' attribute was modified.');
			}
		}
	};
	var observer = new MutationObserver(callback);
	observer.observe(targetNode, config);
	observer.disconnect();
	
onpropertychange
	
	elemToCheck.attachEvent ('onpropertychange', OnAttrModified);   // Internet Explorer
	
mutation observers2

	var observer = new MutationObserver(callback);
	observer.observe(document.documentElement, {
		childList: true,
		subtree: true            
	});
	

what to watch?

	frame.contentDocument
	frame.contentWindow
	document.documentElement
	window.document
	var ifrm = window.frameElement; // reference to iframe element container
	var doc = ifrm.ownerDocument; // reference to container's document
	???
	
	
script injection

	function addScript(JSfileName) {
		var js = document.createElement('script');
		js.setAttribute('type', 'text/javascript');
		js.src = JSfileName;
		childWindowRef.document.body.appendChild(js);
	}
	
	// the injection engine could use inline script instead of a js file
	// doesn't really matter...
	
	// after that a series of eval calls could do the trick

**Comment by inf3rno** [comment]
other things to watch than win.document:

https://developer.mozilla.org/en-US/docs/Web/API/Window
 - history changes
 - pagehide, pageshow events

etc...

According to a dissertation I read the new window runs on a different thread and has a different engine instance, and these engine instances communicate with each other. So I guess there is a limitation about when the child window engine is accessible to the parent window engine. I think this explains why different browsers behave totally different. Still it does not explain why Chrome works differently with and without Karma. I need to open an issue for that in the karma chrome launcher repo.

**Comment by inf3rno** [comment]
As far as I can tell Proxy is just a decorator, so changes won't go through it when we change the property of the original object. We can test it though, but I am pretty sure it works this way.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy

**Comment by inf3rno** [comment]
Yepp, here is an Object.observe with Proxy: https://github.com/gullerya/object-observer It has to convert the original object to a Proxy object, otherwise it won't work.

**Comment by inf3rno** [comment]
Checked object change watchers

Object.observe

	https://stackoverflow.com/a/45112962/607033
	Chrome and Opera supported it, but currently none of the browsers support it
	
Object.watch

	https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/watch
	Firefox supported it prior v58, but currently none of the browsers support it
	
Object.`__defineSetter__`

	win.document

		ie10 -> don't have it
		ch, op, ff -> can't redefine win.document
		ie11, ph -> can redefine win.document, but does not fire on change
		
	win.location
		
		ch, op, ff -> can't redefine win.location
		ie11 -> method not supported
		ph -> can redefine win.location, but does not fire on change
		
	win.location.href
	
		ch, op, ff -> can't redefine win.location.href
		ie11, ph -> can redefine win.location.href, but does not fire on change
		
	win.history.length
		
		no luck
		
Object.defineProperty

	win.document
	
		ch, op, ff -> can't redefine win.document
		ie10, ie11, ph -> can redefine win.document, but does not fire on change
		
	win.location
	
		ch, op, ff -> can't redefine win.location
		ie10, ie11, ph -> can redefine win.location, but does not fire on change
	
	win.location.href
	
		can't redefine or not supported
	
	win.history.length
	
		did not work

**Comment by inf3rno** [comment]
I found by history that `win.history.length` changes. So from the length we will know even with the `after:loaded` how many times the URI changed.

**Comment by inf3rno** [comment]
The history length is interesting. It changes, and some of the browsers have 3 (blank, x, y), while others only 2. It doesn't really make a difference just like others. All we could do is replacing window.location with a proxy in browsers which allow it, but none of them fired on change, which means that they silently failed or the location object was replaced. I think there is nothing we can do.

We could use the history.length to check whether there was a navigation we did not capture because of the huge next tick delay for example in Opera. I need to test for that. If it is not possible to navigate twice before the `after:load` is called, then it will save the day, otherwise I am not sure what I can do about it.

**Comment by inf3rno** [comment]
We need to check the pagehide, pageshow, popstate, hashchange events. Maybe some of them will fire by location change.

**Comment by inf3rno** [comment]
I think I already mentioned that every window runs on a different thread and they share their engine with each other. I used this as an explanation of the delay, but things do not add up. I mean the script in the head of the loaded document can access the engine of the parent window before the delay. Which means in theory the engine of the child window can be shared before this script in the head runs. This is true for Firefox and Karma + Chrome. So still no explanation of why there is a delay in most of the browsers.

**Comment by inf3rno** [comment]
I am pretty sure navigation is possible before our next tick fires. Checking the history length would help to capture multiple location changes initiated by a child script, but we should be aware that hash changes and pushState will change the URI too, not to mention history.back and forward. If I want to use history.length to ensure that only a single location change has occured, then I have to check all of these scenarios, and their effect on the history length. They are easy to follow, because we have events, and the page won't be reloaded at least by some of them.

**Comment by inf3rno** [comment]
I have interesting news:
 - pageshow - something like loaded, but it is called for history back and forward too
 - pagehide - something like beforeunload, but is called for history back and forward too

I am not sure whether pagehide is cancellable, I have to check. Both are supported by all of the browsers (except IE10), so I can use them.

Some other events supported by Firefox, Chrome, maybe Opera, maybe not IE:

```js
window.stop();
// aborts window load, fires when the stop button is pressed while the page is loading
window.onabort = function() {
  alert("Load aborted.");
}

var handle = window.requestIdleCallback(callback[, options]);
// calls the callback when the next time the browser is idle, the same as next tick I think
window.cancelIdleCallback(handle);

window.close();
// closes the child window from the parent, fires when the child window is closed
window.onclose = funcRef;
window.closed;


window.onloadstart
window.onloadend
// fires before and after the page (and all the resources) are loaded
// in Firefox better to call this than onload + next tick, or maybe this + next tick?
```

**Comment by inf3rno** [comment]
In theory Karma served files are not cached, in practice I am not sure. I need to check it.
From nodejs I can limit image streaming, but I need to turn off the static plugin: https://stackoverflow.com/a/6001457/607033 I need this to check whether the onload fires after the resources are loaded, because in some of the browsers this might not be true. It means that I might need to polyfill page load with page load + image load + iframe load + object load, etc... It is important to be consistent about this...

**Comment by inf3rno** [comment]
I need to check the upper events too and I need to check what happens by
 - location.replace
 - location.href change
 - history back, forward, go
 - link.click
 - form.submit
 - window.close

I have to imitate user activity. I think the injection server will be needed on the long run. For pages which don't have the injection server I will be able to write only Chrome bots with security disabled.

**Comment by inf3rno** [comment]
The high level api should be something like:

```js
const results = await user.action(params);
```

the low level api should use navigation events, e.g.

```js
client.action(params);
client.on("actionCompleted", function (err, results){
   // ...
});
```

we could use action id-s:

```js
const actionId = client.action(params);
client.on("actionCompleted:{actionId}", function (err, results){
   // ...
});
```

But it is hard to decide which action was completed, so we might need a scheduler and call the actions in a series. I am not sure because with this approach the params can be promises instead of values and we don't need the async-await syntax.

It is not that hard to write an utility lib, which converts the promises to values and starts the actual process only after everything is done.  Actuall it is just checking which params are promises and use `Promise.all` to wait for them. The callback can be our code, which processes the params.

So to be honest we don't really need async functions into the lib, just promises.

What actions do we have?

 - history actions (back, forward, go)
 - scripted actions (location.href, location.replace, pushState)
 - hyperlink actions (link.follow, form.submit, link.click?)
 - window actions (abort, close, focus/blur?, minimize/maximize?)

Some of these actions will reload the page, others will load it from history and others will change only the URI. It is hard to tell whether we need to distinguish page reloads. I mean for example we can follow a hyperlink by firing the click event, which might prevent reloading the page and load the content with ajax/sockets instead. On the other hand we can emulate turned off javascript and reload the page with the given URI. The latter is problematic, because in most of the browser we cannot prevent running scripts. Only in Firefox and Karma+Chrome can we remove the scripts from the document before it tries to run them (but even that is not sure).

I think the most basic feature is hyperlink actions, so we must implement those first. The others are not that prevalent, so those can wait.

**Comment by inf3rno** [comment]
I think basic high level api should be something like:

```js
let link =  $("#id");
let doc = await user.follow(link);
```

Link following should run the js code of onclick of present and detect whether it should rewrite the location or not, so it should detect "preventDefault" calls somehow. I guess firing a click event won't follow a link on its own, but nothing is impossible. I need to test this.

**Comment by inf3rno** [comment]
Pagehide and pageshow on Firefox (navigation: blank -> x.html -[1500]-> y.html):

```
beforeunload, 87ms, 1, blank
beforeunload + next tick, 91ms, 1, blank
pagehide, 106ms, 1, blank
unload, 106ms, 1, blank
pagehide + next tick, 113ms, <--, 1, x.html
unload + next tick, 113ms, <--, 1, x.html
polling document, 114ms, <--, 1, x.html
polling uri, 114ms, <--, 1, x.html
polling document.readyState, 114ms, <--, 1, x.html
x child script, 118ms, <--, 1, x.html
x child load, 126ms, 1, x.html
pageshow, 126ms, 1, x.html
pageshow + next tick, 129ms, 1, x.html
beforeunload, 1593ms, 1, x.html
beforeunload + next tick, 1595ms, 1, x.html
polling document, 1596ms, 1, x.html
pagehide, 1606ms, 1, x.html
unload, 1606ms, 1, x.html
polling uri, 1613ms, <--, 2, y.html
polling document.readyState, 1613ms, <--, 2, y.html
polling history length, 1613ms, <--, 2, y.html
polling history length, 1614ms, <--, 2, y.html
pagehide + next tick, 1614ms, <--, 2, y.html
unload + next tick, 1614ms, <--, 2, y.html
y child script, 1614ms, <--, 2, y.html
y child load, 1617ms, 2, y.html

```

**Comment by inf3rno** [comment]
Chrome is interesting, `pagehide + next tick` calls when the new document is a loading state. Interestingly unload + next tick is called with complete state. Maybe this will work with manual Chrome too, I have to test it. For Opera, IE11 and PhantomJS it does not make a difference, it is just the same as unload + next tick. IE 10 does not have this event. I need to test Chrome on manual.

**Comment by inf3rno** [comment]
Chrome manual
```
beforeunload 6ms 1 blank
pagehide 16ms 1 blank
unload 16ms 1 blank
x child script 19ms <-- 2 x.html
x child load 19ms 2 x.html
beforeunload + next tick 622ms 2 x.html
polling document 623ms 2 x.html
polling uri 623ms 2 x.html
polling history length 623ms 2 x.html
pagehide + next tick 623ms 2 x.html
unload + next tick 623ms 2 x.html
y child script 1512ms <-- 3 y.html
y child load 1512ms 3 y.html
polling history length 1513ms 3 y.html
end 3006ms 3 y.html
```

Chrome + Karma
```
beforeunload, 9ms, 1, blank
beforeunload + next tick, 12ms, 1, blank
pagehide, 21ms, 1, blank
unload, 22ms, 1, blank
polling document, 25ms, <--, 2, x.html
pagehide + next tick, 25ms, <--, 2, x.html
polling uri, 26ms, <--, 2, x.html
x child script, 28ms, <--, 2, x.html
x child load, 28ms, 2, x.html
polling history length, 28ms, 2, x.html
unload + next tick, 28ms, 2, x.html
polling document.readyState, 1521ms, <--, 3, y.html
polling history length, 1522ms, <--, 3, y.html
y child script, 1523ms, <--, 3, y.html
y child load, 1524ms, 3, y.html
```

**Comment by inf3rno** [comment]
Chrome + Karma is still different than Chrome on manual.

**Comment by inf3rno** [comment]
Okay, let's move on. The only thing we haven't tried is `window.requestIdleCallback`. I checked support, IE 10, IE 11, PhantomJS 2.1 is not supported. Firefox, Opera, Chrome is supported. I'll try whether using this instead of next tick makes a difference.

**Comment by inf3rno** [comment]
The test results:

```
Chrome
beforeunload, 13ms, 1, blank
beforeunload + next tick, 15ms, 1, blank
beforeunload + idle cb, 21ms, 1, blank
pagehide, 39ms, 1, blank
unload, 39ms, 1, blank
pagehide + next tick, 46ms, <--, 2, x.html
unload + next tick, 46ms, <--, 2, x.html
x child script, 46ms, <--, 2, x.html
x child load, 47ms, 2, x.html
pagehide + idle cb, 60ms, 2, x.html
unload + idle cb, 61ms, 2, x.html
y child script, 1519ms, <--, 3, y.html
y child load, 1519ms, 3, y.html

Opera
beforeunload, 16ms, 1, blank
beforeunload + idle cb, 28ms, 1, blank
pagehide, 50ms, 1, blank
unload, 50ms, 1, blank
x child script, 62ms, <--, 2, x.html
x child load, 62ms, 2, x.html
pagehide + idle cb, 75ms, 2, x.html
unload + idle cb, 76ms, 2, x.html
beforeunload + next tick, 248ms, 2, x.html
pagehide + next tick, 248ms, 2, x.html
unload + next tick, 248ms, 2, x.html
y child script, 2262ms, <--, 3, y.html
y child load, 2263ms, 3, y.html

Firefox
beforeunload, 78ms, 1, blank
beforeunload + next tick, 85ms, 1, blank
pagehide, 102ms, 1, blank
unload, 102ms, 1, blank
pagehide + next tick, 120ms, <--, 1, x.html
unload + next tick, 120ms, <--, 1, x.html
x child script, 134ms, <--, 1, x.html
x child load, 139ms, 1, x.html
pageshow, 139ms, 1, x.html
pageshow + next tick, 142ms, 1, x.html
beforeunload + idle cb, 224ms, 1, x.html
pagehide + idle cb, 238ms, 1, x.html
unload + idle cb, 284ms, 1, x.html
pageshow + idle cb, 303ms, 1, x.html
beforeunload, 1581ms, 1, x.html
beforeunload + next tick, 1583ms, 1, x.html
beforeunload + idle cb, 1584ms, 1, x.html
pagehide, 1587ms, 1, x.html
unload, 1588ms, 1, x.html
pagehide + next tick, 1595ms, <--, 2, y.html
unload + next tick, 1595ms, <--, 2, y.html
y child script, 1595ms, <--, 2, y.html
y child load, 1599ms, 2, y.html
pagehide + idle cb, 1604ms, 2, y.html
unload + idle cb, 1615ms, 2, y.html
```

**Comment by inf3rno** [comment]
I checked loadstart and loadend too. They are Firefox only events. They don't work on the window object, the example in the documentation is for images, so I guess it works on images only and the window object doesn't even support it. https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onloadend

**Comment by inf3rno** [comment]
Chrome manual:

```
beforeunload 7ms 1 blank
beforeunload + idle cb 8ms 1 blank
pagehide 12ms 1 blank
unload 12ms 1 blank
x child script 14ms <-- 2 x.html
x child load 15ms 2 x.html
pagehide + idle cb 16ms 2 x.html
unload + idle cb 16ms 2 x.html
beforeunload + next tick 760ms 2 x.html
pagehide + next tick 761ms 2 x.html
unload + next tick 761ms 2 x.html
y child script 1767ms <-- 3 y.html
y child load 1769ms 3 y.html
end 3761ms 3 y.html
```

**Comment by inf3rno** [comment]
The `unload + idle cb` works in Chrome + manual too, which is great again.

**Comment by inf3rno** [comment]
IE 10 and 11 (and Edge too) has a similar method https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate They don't have test delays, so that is not a big problem, but we can use it too, which is great. PhantomJS does not support this, so that will be the only one with the delay.

```
IE 10
beforeunload, 168ms, 1, blank
beforeunload + next tick, 176ms, 1, blank
beforeunload + immediate, 182ms, 1, blank
unload, 1186ms, 1, blank
x child script, 1199ms, <--, 1, x.html
x child load, 1211ms, 1, x.html
unload + immediate, 1218ms, 1, x.html
unload + next tick, 1224ms, 1, x.html
y child script, 1699ms, <--, 2, y.html
y child load, 1710ms, 2, y.html

IE 11
beforeunload, 163ms, 1, blank
beforeunload + next tick, 170ms, 1, blank
beforeunload + immediate, 177ms, 1, blank
pagehide, 1178ms, 1, blank
unload, 1185ms, 1, blank
x child script, 1197ms, <--, 1, x.html
x child load, 1209ms, 1, x.html
pagehide + immediate, 1216ms, 1, x.html
pagehide + next tick, 1222ms, 1, x.html
unload + next tick, 1229ms, 1, x.html
unload + immediate, 1234ms, 1, x.html
y child script, 1694ms, <--, 2, y.html
y child load, 1704ms, 2, y.html
```

**Comment by inf3rno** [comment]
PhantomJS does not support this, but it does not always has a delay, so maybe that is not a problem. IE 10 and IE 11 don't have delays either. At least in these tests. As far as I remember every browser had this delay except Firefox.

**Comment by inf3rno** [comment]
There is a setImmediate polyfill, which we can check: https://github.com/YuzuJS/setImmediate/blob/master/setImmediate.js the requestIdleCallback has polyfills too. I have to check all of them, maybe I find something better than setTimeout 0.

**Comment by inf3rno** [comment]
The timeout is interesting by requestIdleCallback:
```js
requestIdleCallback(processPendingAnalyticsEvents, { timeout: 2000 });
```
I am not sure what it means. In theory it should take at most 2000 msec to call the function. I am not sure because in theory it is called when the browser is idle. Can it force to call it anyways if it is doing a resource heavy task? If so, then I might be able to force Chrome + manual and Opera to call it when the new document is in a loading state. I'll try.

**Comment by inf3rno** [comment]
Well that's interesting. I tried idle callback with timeout: 1.

```
Opera
unload, 29ms, 1, blank
unload + idle 1, 38ms, <--, 2, x.html
x child script, 47ms, <--, 2, x.html
x child load, 48ms, 2, x.html
unload + idle cb, 99ms, 2, x.html
unload + next tick, 819ms, 2, x.html
y child script, 1832ms, <--, 3, y.html
y child load, 1832ms, 3, y.html

Chrome
unload, 59ms, 1, blank
unload + next tick, 66ms, <--, 2, x.html
unload + idle 1, 66ms, <--, 2, x.html
x child script, 66ms, <--, 2, x.html
x child load, 67ms, 2, x.html
unload + idle cb, 67ms, 2, x.html
y child script, 1526ms, <--, 3, y.html
y child load, 1527ms, 3, y.html

Firefox
unload, 96ms, 1, blank
unload + next tick, 104ms, <--, 1, x.html
unload + idle 1, 105ms, <--, 1, x.html
x child script, 119ms, <--, 1, x.html
x child load, 120ms, 1, x.html
unload + idle cb, 229ms, 1, x.html
unload, 1584ms, 1, x.html
unload + next tick, 1590ms, <--, 2, y.html
unload + idle 1, 1591ms, <--, 2, y.html
y child script, 1594ms, <--, 2, y.html
y child load, 1597ms, 2, y.html
unload + idle cb, 1608ms, 2, y.html
```

**Comment by inf3rno** [comment]
That is cool. Now I can inject code in all 3 browsers, which support requestIdleCallback.

**Comment by inf3rno** [comment]
This setImmediate polyfill is interesting: https://github.com/YuzuJS/setImmediate/blob/master/setImmediate.js it uses various methods:

 - setTimeout(fn, 0)
 - window.postMessage + window.onmessage (delivers id)
 - MessageChannel.postMessage + MessageChannel.onmessage (delivers the function itself)
 - appendChild(script) + script.onreadystatechange
 - process.nextTick() (for nodejs)

I think we'll need a map if we want to try out all of these...

**Comment by inf3rno** [comment]
This version of postMessage uses stopPropagation, which is wise I think: https://gist.github.com/stagas/1225404

**Comment by inf3rno** [comment]
- window.requestAnimationFrame - this is roughly the same as `requestIdleCallback(fn, {timeout: 1000/60)`

**Comment by inf3rno** [comment]
I tried out requestAnimationFrame 

```
Chrome
unload, 23ms, 1, blank
unload + next tick, 32ms, <--, 2, x.html
x child script, 34ms, <--, 2, x.html
x child load, 35ms, 2, x.html
y child script, 1528ms, <--, 3, y.html
y child load, 1529ms, 3, y.html

PhantomJS
unload, 316ms, 1, blank
x child script, 317ms, <--, 1, x.html
x child load, 317ms, 1, x.html
unload + frame, 319ms, 1, x.html
unload + next tick, 319ms, 1, x.html
y child script, 1507ms, <--, 2, y.html
y child load, 1508ms, 2, y.html

Firefox
unload, 133ms, 1, blank
unload + next tick, 143ms, <--, 1, x.html
unload + frame, 145ms, <--, 1, x.html
x child script, 159ms, <--, 1, x.html
x child load, 163ms, 1, x.html
unload, 1634ms, 1, x.html
unload + frame, 1641ms, <--, 2, y.html
unload + next tick, 1646ms, <--, 2, y.html
y child script, 1647ms, <--, 2, y.html
y child load, 1651ms, 2, y.html

Opera
unload, 31ms, 1, blank
x child script, 40ms, <--, 2, x.html
x child load, 41ms, 2, x.html
unload + next tick, 208ms, 2, x.html
y child script, 2230ms, <--, 3, y.html
y child load, 2231ms, 3, y.html

IE 11
unload, 1548ms, 1, blank
x child script, 1602ms, <--, 1, x.html
x child load, 1636ms, 1, x.html
unload + next tick, 1648ms, 1, x.html
unload + frame, 1658ms, 1, x.html
y child script, 2067ms, <--, 2, y.html
y child load, 2114ms, 2, y.html

IE 10
unload, 1485ms, 1, blank
x child script, 1525ms, <--, 1, x.html
x child load, 1554ms, 1, x.html
unload + next tick, 1567ms, 1, x.html
unload + frame, 1576ms, 1, x.html
y child script, 2017ms, <--, 2, y.html
y child load, 2044ms, 2, y.html
```

**Comment by inf3rno** [comment]
Chrome and Opera did not support the frame despite that they have the function. I am not sure why...

**Comment by inf3rno** [comment]
We can't use process.nextTick, but we can check next tick polyfills too. 
 - Promise.resolve().then(yourFunction) is another candidate

**Comment by inf3rno** [comment]
Ok, this is the previous version for all browsers, which supports loaded only:

```js

        var callBacks = [];
        var uglyHax = function () {
            var done = function () {
                uglyHax();
                callBacks.forEach(function (cb) {
                    cb();
                });
            };
            win.addEventListener("unload", function unloadListener() {
                win.removeEventListener("unload", unloadListener); // Firefox remembers, other browsers don't
                setTimeout(function () {
                    // IE10, IE11, Opera, PhantomJS, Chrome has a complete new document at this point
                    // Chrome on Karma, Firefox has a loading new document at this point
                    win.document.readyState; // IE10 and IE11 sometimes fails if I don't access it twice, idk. how or why
                    if (win.document.readyState === "complete")
                        done();
                    else
                        win.addEventListener("load", function () {
                            setTimeout(done, 0);
                        });
                }, 0);
            });
        };
        uglyHax();

        callBacks.push(function () {
            con.log("cb", win.location.href, win.document.readyState);
            if (win.location.href !== "http://localhost:9876/base/test/client/y.html")
                win.location.href = "http://localhost:9876/base/test/client/y.html";
            else
                next();
        });
```

I save it here, because I want a clear code.

**Comment by inf3rno** [comment]
https://stackoverflow.com/questions/41289602/add-defer-or-async-attribute-to-dynamically-generated-script-tags-via-javascript

async is enabled by default, and I doubt that IE 10 or 11 support it.
defer can be set

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script
IE 10 and 11 support them

**Comment by inf3rno** [comment]
I tried again, none of them makes a difference.

**Comment by inf3rno** [comment]
The readystatechange method does not work by scripts.

**Comment by inf3rno** [comment]
Appending inline script does not work in PhantomJS either. I mean it runs the script before the new document is created, which is not good for us.

**Comment by inf3rno** [comment]
Both messaging work for Opera. Using idle callback with timeout 0 and 0.1 does not work for Opera. The Promise.resolve is the same as appending the child script.

For Firefox messaging, animation frame work too, the Promise.resolve does not work here either.

**Comment by inf3rno** [comment]
Messaging does not work in IE and PhantomJS, it is called after the child script ran and it has the delay too...

**Comment by inf3rno** [comment]
The summary of the test results:
```
Chrome 68 + Karma
----------------- forced -----------------------
append inline script, 77ms
Promise.resolve, 77ms
-------------- lightly forced ------------------
postMessage, 92ms
MessageChannel, 93ms
setTimeout 0, 93ms
requestIdleCallback 1, 93ms
---------------- delayed -----------------------
requestIdleCallback, 98ms
requestIdleCallback 0.1, 99ms
requestIdleCallback 0, 99ms
requestAnimationFrame did not fire


Opera 54
----------------- forced -----------------------
append inline script, 187ms
Promise.resolve, 187ms
-------------- lightly forced ------------------
postMessage, 191ms
MessageChannel, 191ms
requestIdleCallback 1, 191ms
---------------- delayed -----------------------
requestIdleCallback, 206ms
requestIdleCallback 0.1, 206ms
requestIdleCallback 0, 206ms
setTimeout 0, 322ms
requestAnimationFrame did not fire


PhantomJS 2.1
----------------- forced -----------------------
append inline script, 357ms
-------------- lightly forced ------------------
---------------- delayed -----------------------
postMessage, 359ms
setTimeout 0, 359ms
requestAnimationFrame, 360ms
MessageChannel, 360ms


IE 10
----------------- forced -----------------------
append inline script, 505ms
-------------- lightly forced ------------------
---------------- delayed -----------------------
script readystatechange, 586ms
MessageChannel, 595ms
requestAnimationFrame, 602ms
setTimeout 0, 614ms
setImmediate, 624ms
postMessage, 646ms


IE 11
----------------- forced -----------------------
append inline script, 606ms
-------------- lightly forced ------------------
---------------- delayed -----------------------
MessageChannel, 672ms
postMessage, 679ms
requestAnimationFrame, 686ms
setTimeout 0, 693ms
setImmediate, 699ms


Firefox 61
----------------- forced -----------------------
append inline script, 148ms
Promise.resolve, 148ms
-------------- lightly forced ------------------
setTimeout 0, 157ms
requestIdleCallback 0.1, 157ms
requestIdleCallback 0, 157ms
requestIdleCallback 1, 158ms
postMessage, 158ms
MessageChannel, 159ms
requestAnimationFrame, 159ms
---------------- delayed -----------------------
requestIdleCallback, 280ms
```

**Comment by inf3rno** [comment]
Sadly IE 10, IE 11 and PhantomJS don't have lightly forced solutions, so I am not able to inject scripts by those.

**Comment by inf3rno** [comment]
Here are the defers:

```js
"setTimeout 0": {
	fn: function (cb) {
		setTimeout(cb, 0);
	}
},
"requestIdleCallback": {
	deps: function () {
		return !!window.requestIdleCallback;
	},
	fn: function (cb) {
		window.requestIdleCallback(cb);
	}
},
"requestIdleCallback 1": {
	deps: function () {
		return !!window.requestIdleCallback;
	},
	fn: function (cb) {
		window.requestIdleCallback(cb, {timeout: 1});
	}
},
"requestIdleCallback 0.1": {
	deps: function () {
		return !!window.requestIdleCallback;
	},
	fn: function (cb) {
		window.requestIdleCallback(cb, {timeout: 0.1});
	}
},
"requestIdleCallback 0": {
	deps: function () {
		return !!window.requestIdleCallback;
	},
	fn: function (cb) {
		window.requestIdleCallback(cb, {timeout: 0});
	}
},
"setImmediate": {
	deps: function () {
		return !!window.setImmediate;
	},
	fn: function (cb) {
		window.setImmediate(cb);
	}
},
"requestAnimationFrame": {
	deps: function () {
		return !!window.requestAnimationFrame;
	},
	fn: function (cb) {
		window.requestAnimationFrame(cb);
	}
},
"postMessage": {
	deps: function () {
		return !!window.postMessage;
	},
	fn: function (cb) {
		var mid = "pm." + Math.random();
		var origin = window.location.origin;
		if (!origin)
			origin = window.location.protocol + '//' + window.location.hostname + (window.location.port ? (':' + window.location.port) : '');
		// IE 10 does not support location.origin
		window.addEventListener("message", function handle(evt) {
			if (evt.source !== window || evt.origin !== origin || evt.data !== mid)
				return;
			window.removeEventListener("message", handle);
			evt.stopPropagation();
			cb();
		});
		window.postMessage(mid, origin);
	}
},
"MessageChannel": {
	deps: function () {
		return !!window.MessageChannel;
	},
	fn: function (cb) {
		var channel = new window.MessageChannel();
		channel.port1.onmessage = function (e) {
			cb();
		};
		channel.port2.postMessage(1);
	}
},
"Promise.resolve": {
	deps: function () {
		return !!window.Promise;
	},
	fn: function (cb) {
		window.Promise.resolve().then(cb);
	}
},
"append inline script": {
	fn: function (cb) {
		var body = document.body;
		var script = document.createElement("script");
		script.text = 'window.mycb1();';
		window.mycb1 = function () {
			cb();
			delete window.mycb1;
			body.removeChild(script);
		};
		body.appendChild(script);
	}
},
"script readystatechange": {
	deps: function () {
		return "onreadystatechange" in document.createElement("script");
	},
	fn: function (cb) {
		var body = document.body;
		var script = document.createElement("script");
		script.type = "text/javascript";
		script.src = "javascript:";
		script.onreadystatechange = function () {
			delete script.onreadystatechange;
			body.removeChild(script);
			cb();
		};
		body.appendChild(script);
	}
}
```

**Comment by inf3rno** [comment]
Conclusions:
 - append inline the only forced solution in IE 10, IE 11 and PhantomJS, we can use that to log changes after the unload event but before the new document is created (note that IE has a loading old document, while PhantomJS has a complete old document). Promise.resolve is an alternative for Opera, Chrome, Firefox
 - lightly forced is postMessage, MessageChannel, requestIdleCallback(1) for Opera, Chrome, Firefox, so those can be used in that browsers, while IE 10, IE 11, PhantomJS do not have lightly forced solutions, so they cannot capture the new document in a loading state, just in a complete state. lightly forced is good, because we can replace the tracking document and URI before any script in the loaded page can navigate away or replace the URI. all we could do is adding a script to the head of the page, but that would mean a testing and a non-testing version of the page, and I'd like to avoid that if it is not completely necessary. the injection server will be bad enough...
 - delayed is requestIdleCallback() for Opera, Chrome, Firefox, and setTimeout 0 for IE 10, IE 11, PhantomJS. another option to use a lightly forced solution and add load + defer when it fires, but that is a little bit ugly. this is good for a low quality tracking solution, but the scripts on the loaded page can navigate away before the new document and the URI can be logged. so it will result a gappy tracking history. the main purpose of this should be the notification of the testing script about the loaded page and not the tracking.

Overall, we can add a reliable after-unload and after-load script and a not that reliable loading script.

**Comment by inf3rno** [comment]
Conclusions:
 - append inline the only forced solution in IE 10, IE 11 and PhantomJS, we can use that to log changes after the unload event but before the new document is created (note that IE has a loading old document, while PhantomJS has a complete old document). Promise.resolve is an alternative for Opera, Chrome, Firefox
 - lightly forced is postMessage, MessageChannel, requestIdleCallback(1) for Opera, Chrome, Firefox, so those can be used in that browsers, while IE 10, IE 11, PhantomJS do not have lightly forced solutions, so they cannot capture the new document in a loading state, just in a complete state. lightly forced is good, because we can replace the tracking document and URI before any script in the loaded page can navigate away or replace the URI. all we could do is adding a script to the head of the page, but that would mean a testing and a non-testing version of the page, and I'd like to avoid that if it is not completely necessary. the injection server will be bad enough...
 - delayed is requestIdleCallback() for Opera, Chrome, Firefox, and setTimeout 0 for IE 10, IE 11, PhantomJS. another option to use a lightly forced solution and add load + defer when it fires, but that is a little bit ugly. this is good for a low quality tracking solution, but the scripts on the loaded page can navigate away before the new document and the URI can be logged. so it will result a gappy tracking history. the main purpose of this should be the notification of the testing script about the loaded page and not the tracking.

Overall, we can add a reliable after-unload and after-load script and a not that reliable loading script.

**Comment by inf3rno** [comment]
We need to check the forced solutions whether they are sync or really async. To do that all we need is adding a console.log after the deferAll call and check the order.

**Comment by inf3rno** [comment]
The append inline is sync, so we don't have a forced defer to IE and PhantomJS. All we can do is using an unload event listener. This is a problem because if our event listener is before another event listener which redirects the page, then we won't be able to log that redirection.

**Comment by inf3rno** [comment]
I think I was wrong about the MutationObserver and MutationEvent approaches. I should have added them to the parent document and not to the child document. I have to try them again.

**Comment by inf3rno** [comment]
By requestIdleCallback the 0.1 and the 0 are not valid timeout params. I can clear those tests.

**Comment by inf3rno** [comment]
IE 11 and PhantomJS have MutationObserver support. According to the article I linked, it should be a microtask, so it might be good for nextTick, but firing the event is a task, which makes it bad for nextTick. I have to write a code which utilizes it.

**Comment by inf3rno** [comment]
I tested with MutationObserver, MutationEvent, onpropertychange too and added the append script again. I fixed the timeOrigin too.

```
Chrome manual
---------------------- sync -----------------------
append inline script 1ms
MutationEvent 2ms
---------------------- async ----------------------
Promise.resolve 2ms
MutationObserver 2ms
--------------------- loading ---------------------
postMessage 4ms
MessageChannel 5ms
requestIdleCallback 1 6ms
--------------------- loaded ----------------------
requestIdleCallback 9ms
setTimeout 0 42ms
requestAnimationFrame 1011ms


Chrome Karma
---------------------- sync -----------------------
append inline script 1ms
MutationEvent, 2ms
---------------------- async ----------------------
Promise.resolve, 3ms
MutationObserver, 3ms
--------------------- loading ---------------------
postMessage, 15ms
MessageChannel, 15ms
setTimeout 0, 17ms
requestIdleCallback 1, 17ms
--------------------- loaded ----------------------
requestIdleCallback, 103ms
requestAnimationFrame did not fire



Opera
---------------------- sync -----------------------
append inline script 1ms
MutationEvent, 1ms
---------------------- async ----------------------
Promise.resolve, 2ms
MutationObserver, 2ms
--------------------- loading ---------------------
postMessage, 5ms
MessageChannel, 5ms
requestIdleCallback 1, 5ms
--------------------- loaded ----------------------
setTimeout 0, 296ms
requestAnimationFrame did not fire
requestIdleCallback, 2315ms


IE 11
---------------------- sync -----------------------
append inline script 1ms
MutationEvent, 13ms
---------------------- async ----------------------
--------------------- loading ---------------------
--------------------- loaded ----------------------
MutationObserver, 114ms
MessageChannel, 124ms
requestAnimationFrame, 132ms
setTimeout 0, 143ms
setImmediate, 151ms
postMessage, 158ms


IE 10
---------------------- sync -----------------------
append inline script 1ms
MutationEvent, 6ms
onpropertychange, 15ms
---------------------- async ----------------------
--------------------- loading ---------------------
--------------------- loaded ----------------------
script readystatechange, 79ms
MessageChannel, 89ms
requestAnimationFrame, 96ms
setTimeout 0, 104ms
setImmediate, 114ms
postMessage, 121ms


Firefox
---------------------- sync -----------------------
append inline script 1ms
MutationEvent, 1ms
---------------------- async ----------------------
Promise.resolve, 1ms
MutationObserver, 1ms
--------------------- loading ---------------------
setTimeout 0, 10ms
requestIdleCallback 1, 10ms
postMessage, 11ms
MessageChannel, 12ms
requestAnimationFrame, 14ms
--------------------- loaded ----------------------
requestIdleCallback, 41ms



PhantomJS
---------------------- sync -----------------------
append inline script 1ms
MutationEvent, 1ms
---------------------- async ----------------------
MutationObserver, 1ms
--------------------- loading ---------------------
--------------------- loaded ----------------------
postMessage, 2ms
setTimeout 0, 3ms
requestAnimationFrame, 3ms
MessageChannel, 3ms
```

**Comment by inf3rno** [comment]
Here is the whole testing code. I move on and make a defer.js file based on the results.

```js
describe("xxx", function () {
    it("yyy", function (next) {

        var uuid = "win." + Math.random();
        var timeOrigin;
        var win = window.open("about:blank", uuid, "menubar=yes,location=yes,resizable=yes,scrollbars=yes,status=yes");
        var doc = win.document;
        var con = console;

        console = {
            log: function () {
                var messages = Array.prototype.slice.call(arguments);
                var isDocRenewed = win.document !== doc;
                win.document.readyState; // for IE 10 and 11
                var isLoading = win.document.readyState === "loading";
                if (timeOrigin)
                    messages.push((new Date() - timeOrigin) + "ms");
                if (isDocRenewed && isLoading)
                    messages.push("<-[DL]-");
                else if (isDocRenewed)
                    messages.push("D");
                else if (isLoading)
                    messages.push("L");
                messages.push(win.history.length || 0);
                messages.push(win.location.pathname.slice(win.location.pathname.lastIndexOf("/") + 1));
                con.log.apply(con, messages);
            }
        };

        var defers = {
            "setTimeout 0": {
                fn: function (cb) {
                    setTimeout(cb, 0);
                }
            },
            "requestIdleCallback": {
                deps: function () {
                    return !!window.requestIdleCallback;
                },
                fn: function (cb) {
                    window.requestIdleCallback(cb);
                }
            },
            "requestIdleCallback 1": {
                deps: function () {
                    return !!window.requestIdleCallback;
                },
                fn: function (cb) {
                    window.requestIdleCallback(cb, {timeout: 1});
                }
            },
            "setImmediate": {
                deps: function () {
                    return !!window.setImmediate;
                },
                fn: function (cb) {
                    window.setImmediate(cb);
                }
            },
            "requestAnimationFrame": {
                deps: function () {
                    return !!window.requestAnimationFrame;
                },
                fn: function (cb) {
                    window.requestAnimationFrame(cb);
                }
            },
            "postMessage": {
                deps: function () {
                    return !!window.postMessage;
                },
                fn: function (cb) {
                    var mid = "pm." + Math.random();
                    var origin = window.location.origin;
                    if (!origin)
                        origin = window.location.protocol + '//' + window.location.hostname + (window.location.port ? (':' + window.location.port) : '');
                    // IE 10 does not support location.origin
                    window.addEventListener("message", function handle(evt) {
                        if (evt.source !== window || evt.origin !== origin || evt.data !== mid)
                            return;
                        window.removeEventListener("message", handle);
                        evt.stopPropagation();
                        cb();
                    });
                    window.postMessage(mid, origin);
                }
            },
            "MessageChannel": {
                deps: function () {
                    return !!window.MessageChannel;
                },
                fn: function (cb) {
                    var channel = new window.MessageChannel();
                    channel.port1.onmessage = function (e) {
                        cb();
                    };
                    channel.port2.postMessage(1);
                }
            },
            "Promise.resolve": {
                deps: function () {
                    return !!window.Promise;
                },
                fn: function (cb) {
                    window.Promise.resolve().then(cb);
                }
            },
            "append inline script": {
                fn: function (cb) {
                    var body = document.body;
                    var script = document.createElement("script");
                    script.text = 'window.mycb1();';
                    window.mycb1 = function () {
                        cb();
                        delete window.mycb1;
                        body.removeChild(script);
                    };
                    body.appendChild(script);
                }
            },
            "script readystatechange": {
                deps: function () {
                    return "onreadystatechange" in document.createElement("script");
                },
                fn: function (cb) {
                    var body = document.body;
                    var script = document.createElement("script");
                    script.type = "text/javascript";
                    script.src = "javascript:";
                    script.onreadystatechange = function () {
                        delete script.onreadystatechange;
                        body.removeChild(script);
                        cb();
                    };
                    body.appendChild(script);
                }
            },
            "MutationObserver": {
                deps: function (){
                    return !!window.MutationObserver;
                },
                fn: function (cb){
                    var input = document.createElement("input");
                    input.type = "hidden";
                    var observer = new window.MutationObserver(function () {
                        observer.disconnect();
                        document.body.removeChild(input);
                        cb();
                    });
                    observer.observe(document.body, {
                        childList: true
                    });
                    document.body.appendChild(input);
                }
            },
            "MutationEvent": {
                deps: function (){
                    return !!window.MutationEvent;
                },
                fn: function (cb){
                    var input = document.createElement("input");
                    input.type = "hidden";
                    document.body.addEventListener("DOMNodeInserted", function fn(){
                        document.body.removeEventListener("DOMNodeInserted", fn);
                        document.body.removeChild(input);
                        cb();
                    });
                    document.body.appendChild(input);
                }
            },
            "onpropertychange": {
                deps: function (){
                    return "onpropertychange" in document.documentElement;
                },
                fn: function (cb){
                    var input = document.createElement("input");
                    input.type = "hidden";
                    input.onpropertychange = function (){
                        document.body.removeChild(input);
                        cb();
                    };
                    document.body.appendChild(input);
                    input.value = "trigger";
                }
            }
        };

        var deferAll = function (cb) {
            var registry = {};
            for (var name in defers) {
                var desc = defers[name];
                if (desc.deps && !desc.deps())
                    continue;
                registry[name] = 0;
                desc.fn((function (name) {
                    return function () {
                        registry[name] += 1;
                        cb(name);
                    };
                })(name));
            }
            return registry;
        };

        win.addEventListener("beforeunload", function () {
            console.log("beforeunload");
        });

        win.addEventListener("unload", function () {
            console.log("unload");
            timeOrigin = new Date();
            var reg = deferAll(function (name) {
                console.log(name);
            });
            console.log("unload2");
            setTimeout(function () {
                con.log("# registry", reg);
            }, 1400);
        });

        win.addEventListener("pagehide", function () {
            console.log("pagehide");
        });

        win.addEventListener("pageshow", function () {
            console.log("pageshow");
        });


        win.location.href = "http://localhost:9876/base/test/client/x.html";
        setTimeout(function () {
            win.location.href = "http://localhost:9876/base/test/client/y.html";
        }, 1500);
        setTimeout(next, 2500);
    });
});
```

**Comment by inf3rno** [comment]
I tested Safari too. It does not have an unload event, but the pagehide worked. If we want an universal solution, then we have to address that. I need to check whether I can use pagehide or beforeunload instead of unload.

**Comment by inf3rno** [comment]
If the async init is not mandatory e.g. blank has to load first, then I think it is better to test it by the first usage. So lazy init is better than async init.

**Comment by inf3rno** [comment]
I'll remove my defer abstractions and try different defer solutions with lazy init.

**Comment by inf3rno** [comment]
I think we can close this. We can detect page loads with beforeunload/unload/pagehide + defers. By initiating the first page load we can try out which approaches work.

**Comment by inf3rno** [comment]
Was a hard task.
